### ISSUE TITLE
[BUG] Network graph auto scale clipping

### ISSUE THREAD

**Issue by RedBearAK** [aristocratos/bpytop#257]
bpytop: 1.0.61
psutil: 5.8.0
Linux Mint 20.1 (64-bit)
Acer Aspire 5 Ryzen 3500U

I frequently have observed the network graph appear to be "clipping", often by quite large amounts. This is not the expected behavior while "auto" scaling is enabled. See screenshots: 

![Network_auto_scale_clipping](https://user-images.githubusercontent.com/64876997/106686930-bc52f900-6577-11eb-9faa-c92e18b1cbb3.png)

![Network_auto_scale_clipping2](https://user-images.githubusercontent.com/64876997/106687470-a42fa980-6578-11eb-8d66-7b883d415dd5.png)

![Network_auto_scale_clipping3](https://user-images.githubusercontent.com/64876997/106687476-a4c84000-6578-11eb-87cb-84ac1e62ec58.png)

Expected behavior is to always be able to see the tops of the highest columns of data displayed on the graph, if "auto" scaling is turned on. 

Like this: 

![Network_auto_scale_no_clipping](https://user-images.githubusercontent.com/64876997/106687499-b1e52f00-6578-11eb-9811-6e53e11dbc6b.png)

This has been the observed behavior for me of the network graph auto-scaling during all the time (almost a year) and on all the different machines on which I've used BPyTOP. I've tried many times switching "auto" on and off and nothing seems to make it more reliable at showing the entire graph. 

I tried in the past to examine the code to find the algorithm that is responsible for the auto-scaling. Due to my poor Python reading skills I wasn't very successful. If you can point me to the exact code section I can at least take a look at it and try some experimental changes.

**Comment by aristocratos** [comment]
@RedBearAK 

> I tried in the past to examine the code to find the algorithm that is responsible for the auto-scaling.

That would be the following lines:

https://github.com/aristocratos/bpytop/blob/2fbdc35125ad5d9a9f613736f194a33ec69bb14b/bpytop.py#L3589-L3605

On possibility for the autoscaling not working well could be a very low update time for bpytop, like sub 1000ms. Since the net speed stats might "stutter" then, showing 100mbs first update, 30kbs next update and so on. This would cause the trigger for rescaling the graph to be delayed since the first update would add to the `stat["graph_raise"]` counter and the second update would subtract.

Can you post your `bpytop.conf` just to see if there is any weird combination of options that's causing this if you're not using a very low update counter.

**Comment by RedBearAK** [comment]
@aristocratos 

Well, you know me, I do use low update times for fun sometimes, like to fill up the graph quickly, but here is another screenshot with the update having been set to 1500ms for many hours. 

![Network_auto_scale_clipping_fullwindow](https://user-images.githubusercontent.com/64876997/106998973-d59fa500-6731-11eb-930a-55ac6766c55f.png)

Doubt the `bpytop.conf` will help but here it is below. As far as I am aware I've seen this behavior on multiple different machines over the past year. The scale just keeps jumping around and never seems to reliably keep from clipping. 

[bpytop.conf.txt](https://github.com/aristocratos/bpytop/files/5930631/bpytop.conf.txt)

**Comment by aristocratos** [comment]
@RedBearAK 
> The scale just keeps jumping around and never seems to reliably keep from clipping.

That's how it's supposed to work. It will auto adjust the current scale to the last 5 (consistently) measured speeds. In your last screenshot the max value for graph is at 10K, so any values above 10K will clip.

If you don't want any clipping you need to turn of auto scaling and set the manual values to your max speed.

**Comment by RedBearAK** [comment]
@aristocratos 

Ah, I see. I expected that auto-scaling meant that every bit of data being displayed on the graph would be included in the calculation for the current scale. That explains why I never understood what it was doing. And why it's so jumpy. 

Do you consider it too difficult to keep track of the max value for the entire displayed width of the graph? Or at least the newest half or two-thirds of the measured values?

**Comment by aristocratos** [comment]
@RedBearAK 

> Do you consider it too difficult to keep track of the max value for the entire displayed width of the graph? Or at least the newest half or two-thirds of the measured values?

Not at all, the max value is based on values from the data set which holds all the values currently seen on the graph. But the downside of increasing the scope for the max value is that small values won't show up on the graph at all if they are below 1% of the current max value. For example if the current max value is 10MiB, any value below 100KiB will not be shown and increasing the range of the shown values isn't very practical either since the size of the net box is usually pretty small.

**Comment by RedBearAK** [comment]
@aristocratos 

>Not at all, the max value is based on values from the data set which holds all the values currently seen on the graph. But the downside of increasing the scope for the max value is that small values won't show up on the graph at all if they are below 1% of the current max value. For example if the current max value is 10MiB, any value below 100KiB will not be shown and increasing the range of the shown values isn't very practical either since the size of the net box is usually pretty small.

You know, I've actually been thinking about that for quite a while. I have noticed what you're describing, that if there is a huge spike in network activity it can make the slower activity completely disappear because it goes below the minimum value of "one dot". I can't help but think it might be good to force the graph to show at least one visible dot if there is any activity at all. Sort of set a "round up to 1 dot" minimum limit unless the measured activity is actually "zero". I feel like a lot of users would immediately recognize the value in that without even needing to explain it. A gap in the graph should intuitively correspond to "no meaningful network activity at all, not even a single normal 1500 byte packet" rather than "yeah, there was 900K of network activity but you're saturating a 100Gbps connection now so I'm going to pretend there was no activity". 

In addition I've been thinking the highly variable scale of the network graph might benefit from moving to a logarithmic scale, rather than a linear scale. It's already a "heat" graph in the sense that the color changes as the columns get taller, just like the CPU graph. What if those color changes (somewhat) corresponded to powers-of-10 changes in the scale? 

I feel like that would make the network graph take much better advantage of the tiny box it's usually stuck inside. It would also drastically decrease the frequency of the need to change the auto scaling, since it could show both very small and very large values at the same time. Even at the smallest box size, which yields only two "character heights", it would be possible to go from one dot = 1Kbps to eight dots = 10Gbps. Or one dot = 10bps to eight dots = 100Mbps, as a minimum scale. That's quite the range. 

So that's the dream I'm having now. Logarithmic network graph scale with a one-dot minimum for any activity of at least 10bps (customizable by the user). 

Oh, wow, you decreased the minimum window size to 60x24 when the `proc` box isn't displayed. That's impressive. And the way the user can just freely hide any box they want now with the 1-2-3-4 keys is a genius feature. It's become a highly adaptable application.

**Comment by aristocratos** [comment]
@RedBearAK 
I've changed so any value above 0 always register on graph in v1.0.62

Adding code for switching to a logarithmic scale for the network graph (would also be useful for the graphs in IO mode) is a good idea and will look into it when I've got time. 

Also, shouldn't be a problem adding some options for manually setting the sample size for the autoscaling feature.

> Oh, wow, you decreased the minimum window size to 60x24 when the proc box isn't displayed.

The required size will adapt to the boxes currently shown, so you can do stuff like this:
![Screenshot 2021-02-14 at 11 37 49](https://user-images.githubusercontent.com/59659483/107874348-3c068f00-6eb9-11eb-8da8-8b1976e6f0de.png)

:)

**Comment by RedBearAK** [comment]
@aristocratos 

Love it. So cool. How are you doing those frameless windows? Using i3 or something? 

The one-dot-minimum change looks very good. However with themes that use very similar colors for "up" and "down" it can have a tendency to blend in with the other side of the graph. Making it appear as if there is "nothing" on the other side sometimes, when there isn't enough traffic to make any "blips" on the other side. 

I wonder if it might be wise to... and I just realized there's probably no good way to insert any kind of thin line or space between the two sides since it's a character-based display. The app is so slick I keep forgetting it's just pretending to be a graphical display. Unless... is there a set of those dot-matrix characters with a solid line on top, or bottom? But I guess that would just end up looking like a dashed line if the font rendering is off. Still, it could work as a visual separator. Probably a pain to even attempt to implement though, using a different set of characters for the center row(s). 

FYI, I'm still seeing the lines around the battery info widget lagging behind when switching themes. Looks like changing the theme is still not triggering an immediate redraw of the battery widget.

**Comment by RedBearAK** [comment]
@aristocratos 

I could swear I was seeing the network graph in 1.0.62 showing at least one dot all the way across the graph even when the traffic (upload) was often at a flat “0 bitps”. But in 1.0.63 that’s no longer the case. So if you tweaked something there it was a good change. I’m looking at a graph now that is frequently showing just 3 or 4 solitary dots on the upload side because there’s so little upload traffic. It feels very easy to understand what it’s telling me.

**Comment by aristocratos** [comment]
@RedBearAK 
> Using i3 or something?

No, it's iTerm2 on OsX.

> However with themes that use very similar colors for "up" and "down" it can have a tendency to blend in with the other side of the graph. 

Well, that's why the upload and download graphs have separate color options, if the theme you are using have the same colors for upload and download and it's a problem, edit the theme.
Inserting a spacer between them, which would take up a separate line, would not be a good solution when there are better options.

> I could swear I was seeing the network graph in 1.0.62 showing at least one dot...

Ye, fixed some rounding errors.

**Comment by RedBearAK** [comment]
@aristocratos 

> No, it's iTerm2 on OsX.

I be darned. Didn’t know iTerm2 could do that. Thanks. 

The code tweak from fixing the rounding error has made it highly unlikely the user would ever again see just a solid line of single dots underneath the download graph, so the whole “blending colors” illusion isn’t really an issue anymore. As long as there are a few “gaps” or “blips” down there it becomes much more visually obvious that the two sides are separate. So no worries. You waved your magic developer wand and the problem vanished.
